### PR TITLE
Added basic map API

### PR DESCRIPTION
**Pull request by Creeperface01**
I also added some map saving, I know its bad, but idk how to save in, because map data are long array and we need byte array. But it works. The only problem is when player take the map item from item frame and pickup it, because it throws EOF exception, idk if maybe player sends bad comppound tag

### CODE REVIEW

**Comment by NycuRO** [comment]
👍

**Comment by kazuemon** [comment]
nice :)

**Comment by PikyCZ** [comment]
woow :)

File: src/main/java/cn/nukkit/item/ItemMap.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+            if (img.getHeight() != 128 || img.getWidth() != 128) { //resize
+                image = new BufferedImage(128, 128, img.getType());
+                Graphics2D g = image.createGraphics();
+                g.drawImage(img, 0, 0, 128, 128, null);
+                g.dispose();
+            }
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            ImageIO.write(image, "png", baos);
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/item/ItemMap.java]: @Creeperface01 Hmm.... you can add and for video like MiNET?

And for add image "png", replace "image" with name or..?

Comment (Creeperface01) [inline src/main/java/cn/nukkit/item/ItemMap.java]: what :D

Comment (Creeperface01) [inline src/main/java/cn/nukkit/item/ItemMap.java]: and I think video should be added by plugin

Comment (NycuRO) [inline src/main/java/cn/nukkit/item/ItemMap.java]: You can add support and for video? Like MiNET?

And how add images with that? 📦

Comment (NycuRO) [inline src/main/java/cn/nukkit/item/ItemMap.java]: Ok, exampleplugin please? ⭕️

Comment (vzlomka76) [inline src/main/java/cn/nukkit/item/ItemMap.java]: @NycuRO it's easy, please use gif. that get a card with a picture please use:
ItemMap map = new ItemMap();
map.setImage(ImageIO.read(filewithimage));
player.getInventory.addItem(map);

Comment (NycuRO) [inline src/main/java/cn/nukkit/item/ItemMap.java]: @vzlomka76 Ok, and what event need use?

And in "filewithimage" need add directory ? Example: /root/server/image.gif?

**Comment by Creeperface01** [comment]
Of course

File: src/main/java/cn/nukkit/item/ItemMap.java (hunk 1 above)
Comment (YaBoiBalAnce) [inline src/main/java/cn/nukkit/item/ItemMap.java]: `itemMap.setImage(ImageIO.read(new File(getDataFolder(),"banner.png")));` 
This worked for me

**Comment by kazuemon** [comment]
11:54:04 [ALERT] java.lang.RuntimeException: java.io.EOFException
        at cn.nukkit.item.Item.parseCompoundTag(Item.java:36)
        at cn.nukkit.item.Item.getNamedTag(Item.java:1804)
        at cn.nukkit.item.Item.deepEquals(Item.java:1998)
        at cn.nukkit.item.Item.deepEquals(Item.java:1991)
        at cn.nukkit.item.Item.deepEquals(Item.java:1987)
        at cn.nukkit.Player.handleDataPacket(Player.java:2187)
        at cn.nukkit.network.Network.processPackets(Network.java:180)
        at cn.nukkit.network.Network.processBatch(Network.java:156)
        at cn.nukkit.Player.handleDataPacket(Player.java:1895)
        at cn.nukkit.network.RakNetInterface.handleEncapsulated(RakNetInterface
java:155)
        at cn.nukkit.raknet.server.ServerHandler.handlePacket(ServerHandler.jav
:123)
        at cn.nukkit.network.RakNetInterface.process(RakNetInterface.java:66)
        at cn.nukkit.network.Network.processInterfaces(Network.java:76)
        at cn.nukkit.Server.tick(Server.java:1010)
        at cn.nukkit.Server.tickProcessor(Server.java:808)
        at cn.nukkit.Server.start(Server.java:787)
        at cn.nukkit.Server.<init>(Server.java:472)
        at cn.nukkit.Nukkit.main(Nukkit.java:68)
Caused by: java.io.EOFException
        at java.io.DataInputStream.readFully(Unknown Source)
        at java.io.DataInputStream.readFully(Unknown Source)
        at cn.nukkit.nbt.stream.NBTInputStream.readFully(NBTInputStream.java:45

        at cn.nukkit.nbt.tag.ByteArrayTag.load(ByteArrayTag.java:36)
        at cn.nukkit.nbt.tag.Tag.readNamedTag(Tag.java:88)
        at cn.nukkit.nbt.tag.CompoundTag.load(CompoundTag.java:35)
        at cn.nukkit.nbt.tag.Tag.readNamedTag(Tag.java:88)
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:77)
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:72)
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:90)
        at cn.nukkit.item.Item.parseCompoundTag(Item.java:34)
        ... 17 more

**Comment by Creeperface01** [comment]
Maps are not for normal use yet.

**Comment by Snake1999** [comment]
For my own opinion on video: @Creeperface01 @NycuRO 
1. Videos are really a good function, but adding all what we thought is useful will make nukkit more obese and complex rather than simple. It should be added as an extra util instead of what is pushed to the core.
2. All in all we should build the API what is EASY TO USE by developers. That means we shouldn't push complex videos and take long time doing bug-fixing ,issue-answering, and complain-suffering. We should concentrate on more fantastic not flaring functions.

**Comment by Creeperface01** [comment]
anyone has an better idea how to save an image? :D

**Comment by Tee7even** [comment]
As for videos, it's done by just changing images really frequently. Not really "videos", more like simple animations. And I wouldn't recommend actually using it on servers.

**Comment by NycuRO** [comment]
@Tee7even Sorry, but it's very complicated for starters on java and nukkit. All not using nukkit because not exist wiki, not exist plugins very goods and others...

For example on me: I want to use nukkit but i don't use because i don't know how create plugins and more like pocketmine.

I want add maps to server, but i don't know how.

**Comment by Tee7even** [comment]
@NycuRO Anyway you won't succeed in using it by begging people to get something done for you. Nukkit is such server software that doesn't have an active plugin developer community. Just because it doesn't and it won't. Most of the people here have their own servers, **they don't care about _your_ needs.** The most suitable way to use Nukkit is to code plugins by yourself. Learn to code, @NycuRO. And English, really difficult to understand you sometimes.

**Comment by Creeperface01** [comment]
I learned php and pocketmine API quite well, then i found nukkit, beginning was hard because i didnt know java but i learned quite fast and i didnt have any problem with nukkit API because its very similar to pocketmine

**Comment by YaBoiBalAnce** [comment]
I found running gifs quite easy. this however was using multiple images to work as "frames"'
giphy

**Comment by Tee7even** [comment]
>this however was using multiple images to work as "frames"'

It is done exactly like that, there's no other way to do it.

**Comment by sero583** [comment]
@Tee7even or looping a video to single image's?

**Pull request merged**